### PR TITLE
[inductor][cpp] support bf16/fp16 gemm template epilogue fusion

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127195
* #127505
* #127687
* __->__ #126545

As part of #125683, this PR adds epilogue fusion support for bf16/fp16 gemms. The key changes are as follows:
1. bf16 linear w/ epilogue fusion of some ops was originally supported via ATen oneDNN linear pointwise ops. In order to match the ATen op semantics, in-template epilogue support is added to the cpp gemm template so that we would have: "gemm + in-template epilogues -> template buffer". If the template is chosen for codegen, the in-template epilogues will be concatenated with the out-of-template epilogues that are appended during the scheduling.
2. Support bf16/fp16 legalization for `codegen_loop_bodies` which is used to generate the epilogue loops.
3. We used to leverage the in-place buffer mechanism to handle the in-place buffers in the epilogue codegen, in particular, for the reuses for output buffers of GEMM, template and epilogues. This is not correct since the output buffer is an "output" not an "in-place" buffer of the template kernel itself. Now, we use a dedicated "aliases" dict to manage such buffer reuses and the intermediate aliasing buffers are removed after codegen.
4. Add `localize_buffer` method to `LocalBufferScope` to allow the replacement of a global buffer with a local one in the given inductor IR nodes. This helps the fused loops to work on smaller-sized local buffers for better data locality.

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 1]
```diff
@@ -217,26 +222,30 @@ def add_choices(
         input_nodes,
         beta=1,
         alpha=1,
+        has_bias=False,
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:225]: Originally we use the number of input nodes to decide whether there is a bias (2: no bias, 3: with bias) but with the inputs from epilogue nodes as part of the template, there could be more inputs even if there is no bias. So we now use a dedicated flag to check that.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -217,26 +222,30 @@ def add_choices(
...
         trans_w=False,
         input_indices=None,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:228]: It is used to create the in-template epilogue nodes.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 3]
```diff
@@ -395,11 +406,26 @@ def render(  # type: ignore[override]
             Y = template_buffer_node
 
         template_buffer = Y
+        gemm_output_buffer = template_buffer
+
+        epilogues: List[ir.IRNode] = []
+        if self.epilogue_creator is not None:
+            gemm_output_name = "GemmOut"
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:413]: with in-template epilogue nodes, the gemm output could be different from the template output, i.e., gemm out -> in-template epilogues -> template output -> fused out-of-template epilogues. So we create a dedicated buffer for gemm out.

**Comment by jgong5** [comment]
@jansel @lezcano @peterbell10 The dependent PRs have now landed. Could you please take a look at this one? Thanks a lot!

**Review by jansel — APPROVED** [review]

**Comment by henrylhtsang** [comment]
Hi @jgong5, I am debugging the cpu selection algorithm tests, which are sometimes flaky and would fail due to counters["inductor"]["cpp_epilogue_fusion_counter"] = 1. Want to see if you have insights. I don't have a stable repro. It is occasionally failing (say 25%) in fbcode CI.

In particular, I am mostly looking test_linear_with_pointwise. 
* In the 156 tests that are from test_linear_with_pointwise (excluding the test_linear_with_pointwise_dynamic_shapes), 41 of them are flaky.
* If I limit to those that contains bfloat16 in their names, there are 52 tests and 41 of them are flaky (i.e., all flaky tests have bfloat16 in their names)
* If I further limit to those that contains test_linear_with_pointwise_batch_size_384_in_features_196_out_features_385, among the 26 tests, 21 are flaky. 
* If I further limit to "bias_True", then 11 out of 13 tests are flaky. The only two that are not flay have epilogue div or mul.

The other thing I observed is that when the tests pass, they print 
```
AUTOTUNE linear_unary(384x196, 384x196)
72  cpp_packed_gemm_0 0.2068 ms 100.0%
73  _linear_pointwise 236.3641 ms 0.1%
```
but when they failed due to flakiness, they would print 
```
AUTOTUNE mm(384x196, 196x384)
71  cpp_packed_gemm_0 4.1393 ms 100.0%
72  mm 22332.3400 ms 0.0%
```

curious if you have any idea about it.

**Comment by jgong5** [comment]
> Hi @jgong5, I am debugging the cpu selection algorithm tests, which are sometimes flaky and would fail due to counters["inductor"]["cpp_epilogue_fusion_counter"] = 1. Want to see if you have insights. I don't have a stable repro. It is occasionally failing (say 25%) in fbcode CI.
> 
> In particular, I am mostly looking test_linear_with_pointwise.
> 
> * In the 156 tests that are from test_linear_with_pointwise (excluding the test_linear_with_pointwise_dynamic_shapes), 41 of them are flaky.
> * If I limit to those that contains bfloat16 in their names, there are 52 tests and 41 of them are flaky (i.e., all flaky tests have bfloat16 in their names)
> * If I further limit to those that contains test_linear_with_pointwise_batch_size_384_in_features_196_out_features_385, among the 26 tests, 21 are flaky.
> * If I further limit to "bias_True", then 11 out of 13 tests are flaky. The only two that are not flay have epilogue div or mul.
> 
> The other thing I observed is that when the tests pass, they print
> 
> ```
> AUTOTUNE linear_unary(384x196, 384x196)
> 72  cpp_packed_gemm_0 0.2068 ms 100.0%
> 73  _linear_pointwise 236.3641 ms 0.1%
> ```
> 
> but when they failed due to flakiness, they would print
> 
> ```
> AUTOTUNE mm(384x196, 196x384)
> 71  cpp_packed_gemm_0 4.1393 ms 100.0%
> 72  mm 22332.3400 ms 0.0%
> ```
> 
> curious if you have any idea about it.

Hi @henrylhtsang From the log you shared, it seems the `aten.mm` is not replaced by `linear_pointwise` when the tests fail. There is an FX graph pass that does the graph rewrite here: https://github.com/pytorch/pytorch/blob/09f9c256ada73529791f3107b3499339ace29de4/torch/_inductor/fx_passes/mkldnn_fusion.py#L1136

The pass is executed when the `_is_packable_linear` extra checker returns `True`. Not sure what happened when the tests fail. But perhaps you can firstly check if the problem is caused by `_is_packable_linear` returning `False` in the first place and then dig into the checker to understand why? You can also compare the `fx_graph_readable.py` between the successful and failing cases generated with `TORCH_COMPILE_DEBUG=1`. Below is what I got from my side with `TORCH_COMPILE_DEBUG=1 pytest -vsk test_linear_with_pointwise_batch_size_384_in_features_196_out_features_385_bias_True_epilogue_tanh_cpu_bfloat16 test_cpu_select_algorithm.py`. FYI.
```python
class <lambda>(torch.nn.Module):
    def forward(self, arg2_1: "bf16[384, 196]"):
        # No stacktrace found for following nodes
        _frozen_param1: "bf16[385]" = self._frozen_param1
        _frozen_param3 = self._frozen_param3
        
         # File: /home/jgong5/pytorch/test/inductor/test_cpu_select_algorithm.py:236 in forward, code: return self.epilogue(self.linear(x))
        _linear_pointwise_default: "bf16[384, 385]" = torch.ops.mkldnn._linear_pointwise.default(arg2_1, _frozen_param3, _frozen_param1, 'none', [], '');  arg2_1 = _frozen_param3 = _frozen_param1 = None
        tanh: "bf16[384, 385]" = torch.ops.aten.tanh.default(_linear_pointwise_default);  _linear_pointwise_default = None
        return (tanh,)
```

**Comment by henrylhtsang** [comment]
> > Hi @jgong5, I am debugging the cpu selection algorithm tests, which are sometimes flaky and would fail due to counters["inductor"]["cpp_epilogue_fusion_counter"] = 1. Want to see if you have insights. I don't have a stable repro. It is occasionally failing (say 25%) in fbcode CI.
> > In particular, I am mostly looking test_linear_with_pointwise.
> > 
> > * In the 156 tests that are from test_linear_with_pointwise (excluding the test_linear_with_pointwise_dynamic_shapes), 41 of them are flaky.
> > * If I limit to those that contains bfloat16 in their names, there are 52 tests and 41 of them are flaky (i.e., all flaky tests have bfloat16 in their names)
> > * If I further limit to those that contains test_linear_with_pointwise_batch_size_384_in_features_196_out_features_385, among the 26 tests, 21 are flaky.
> > * If I further limit to "bias_True", then 11 out of 13 tests are flaky. The only two that are not flay have epilogue div or mul.
> > 
> > The other thing I observed is that when the tests pass, they print
> > ```
> > AUTOTUNE linear_unary(384x196, 384x196)
> > 72  cpp_packed_gemm_0 0.2068 ms 100.0%
> > 73  _linear_pointwise 236.3641 ms 0.1%
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but when they failed due to flakiness, they would print
> > ```
> > AUTOTUNE mm(384x196, 196x384)
> > 71  cpp_packed_gemm_0 4.1393 ms 100.0%
> > 72  mm 22332.3400 ms 0.0%
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > curious if you have any idea about it.
> 
> Hi @henrylhtsang From the log you shared, it seems the `aten.mm` is not replaced by `linear_pointwise` when the tests fail. There is an FX graph pass that does the graph rewrite here:
> 
> https://github.com/pytorch/pytorch/blob/09f9c256ada73529791f3107b3499339ace29de4/torch/_inductor/fx_passes/mkldnn_fusion.py#L1136
> 
> The pass is executed when the `_is_packable_linear` extra checker returns `True`. Not sure what happened when the tests fail. But perhaps you can firstly check if the problem is caused by `_is_packable_linear` returning `False` in the first place and then dig into the checker to understand why? You can also compare the `fx_graph_readable.py` between the successful and failing cases generated with `TORCH_COMPILE_DEBUG=1`. Below is what I got from my side with `TORCH_COMPILE_DEBUG=1 pytest -vsk test_linear_with_pointwise_batch_size_384_in_features_196_out_features_385_bias_True_epilogue_tanh_cpu_bfloat16 test_cpu_select_algorithm.py`. FYI.
> 
> ```python
> class <lambda>(torch.nn.Module):
>     def forward(self, arg2_1: "bf16[384, 196]"):
>         # No stacktrace found for following nodes
>         _frozen_param1: "bf16[385]" = self._frozen_param1
>         _frozen_param3 = self._frozen_param3
>         
>          # File: /home/jgong5/pytorch/test/inductor/test_cpu_select_algorithm.py:236 in forward, code: return self.epilogue(self.linear(x))
>         _linear_pointwise_default: "bf16[384, 385]" = torch.ops.mkldnn._linear_pointwise.default(arg2_1, _frozen_param3, _frozen_param1, 'none', [], '');  arg2_1 = _frozen_param3 = _frozen_param1 = None
>         tanh: "bf16[384, 385]" = torch.ops.aten.tanh.default(_linear_pointwise_default);  _linear_pointwise_default = None
>         return (tanh,)
> ```

Thanks. Unfortunately I don't have a stable repro, so I might have to land some debugging code to check it first. I plan on checking if `torch.ops.mkldnn._is_mkldnn_bf16_supported()` is False. Any other flags I should pay attention to?

**Comment by jgong5** [comment]
> Thanks. Unfortunately I don't have a stable repro, so I might have to land some debugging code to check it first. I plan on checking if `torch.ops.mkldnn._is_mkldnn_bf16_supported()` is False. Any other flags I should pay attention to?

If that is for debugging purpose only and temporary changes, perhaps you can add some logging info inside `_is_packable_linear` for all the relevant paths that return `False`. This can expediate your debugging I guess.

**Pull request closed without merging**